[10.0][ADD] sale_procurement_amendment

rousseldenis opened the pull request:
Waiting for this change has been merged in 10.0 :
- [x] https://github.com/odoo/odoo/pull/20874. Done in https://github.com/odoo/odoo/commit/58e85bdec08f6f07afe7b8cdfaea6f7cef2776d1

rousseldenis committed:
[10.0][ADD] sale_procurement_amendment

andreparames commented on sale_procurement_amendment/models/sale_order_line.py:
```diff
@@ -0,0 +1,81 @@
+        'procurement_ids',
+        'procurement_ids.state',
+        'procurement_ids.product_qty')
+    def _compute_to_be_procured(self):
```
Should this depend on `product_uom_qty`, or is it not needed?

rousseldenis replied:
Of course!

andreparames commented:
LGTM :+1:

rousseldenis committed:
[IMP] sale_procurement_amendment: Improve README

rousseldenis committed:
[FIX] Don't take into account cancel moves

LoisRForgeFlow commented on sale_procurement_amendment/models/sale_order_line.py:
```diff
@@ -0,0 +1,82 @@
+        if 'product_uom_qty' in values:
+            # Quantities has changed
+            # Check if procurement has to be created
+            for line in lines_lower:
```
Have you considered chained moves and MTO environments? the quantity change has to propagate throughout the chain.

rousseldenis replied:
@lreficent I've added tests to enlight MTO process.

I will add multi procurement rules chained flow tests in a further commit but already tested on an environment with such configuration.

I have to mention that I do not depends on purchase not to pull too much modules, but with the standard process, draft purchases orders are correctly updated as you increase and decrease sale order lines quantities back and forth indefinitely!

rousseldenis replied:
Tests added

LoisRForgeFlow reviewed (changes requested):
Check travis, tests are failing

EDIT: nevermind, I guess because of the odoo PR, right?

rafaelbn commented:
@rousseldenis this PR is for v10 and https://github.com/odoo/odoo/pull/20874 is for v9. Shouldn't be both same version?

cc @carlosdauden

rousseldenis commented:
@rafaelbn Yes, so waiting forwardport after...

rousseldenis committed:
[FIX] sale_procurement_amendment: Cancelled pickings + Add tests.

Already cancelled pickings were not taken into account for pickings in progress computation.
Add tests to highlight MTO processes

rousseldenis committed:
[FIX] sale_procurement_amendment: Dont cancel all procurements at once

When there are several procurements concerning one sale order line, cancelling
them at once (when lowering quantities) confuses Odoo. Cancel them one by one

rousseldenis commented:
@rafaelbn Travis goes green for tests. https://github.com/odoo/odoo/commit/58e85bdec08f6f07afe7b8cdfaea6f7cef2776d1 is on v10 branch now.

rousseldenis commented:
@lreficent Could you update your comments ?
@rafaelbn

nikul-serpentcs commented on sale_procurement_amendment/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+    'version': '10.0.1.0.0',
+    'license': 'AGPL-3',
+    'author': 'ACSONE SA/NV,Odoo Community Association (OCA)',
+    'website': 'https://acsone.eu',
```
@rousseldenis URL must be [this](https://github.com/OCA/maintainer-tools/blob/master/template/module/__manifest__.py#L11)

rousseldenis replied:
Sure

nikul-serpentcs commented on sale_procurement_amendment/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+    'author': 'ACSONE SA/NV,Odoo Community Association (OCA)',
+    'website': 'https://acsone.eu',
+    'depends': [
+        'sale',
```
@rousseldenis Please remove it, Because `sale_stock` is depends `sale`

rousseldenis replied:
Mmmh, will see. As I'd prefer explicit dependances declaration than implicit.

nikul-serpentcs commented on sale_procurement_amendment/models/sale_order_line.py:
```diff
@@ -0,0 +1,86 @@
+
+    to_be_procured = fields.Boolean(
+        compute='_compute_to_be_procured',
+        readonly=True,
```
@rousseldenis remove it, compute always `readonly`

rousseldenis replied:
yes

nikul-serpentcs commented on sale_procurement_amendment/models/stock_move.py:
```diff
@@ -0,0 +1,37 @@
+
+    can_be_amended = fields.Boolean(
+        compute='_compute_can_be_amended',
+        readonly=True,
```
same here

nikul-serpentcs reviewed (changes requested):
Improve Code

rousseldenis committed:
[10.0][IMP] sale_procurement_amendment: remove readonly + test outgoing move qty

rousseldenis commented:
@bodedra @lreficent Could you update your reviews ? Thanks

rousseldenis commented:
@Cedric-Pigeon

rousseldenis committed:
[10.0][IMP] sale_procurement_amendment: Add README fragments

nikul-serpentcs reviewed (approved):
@rousseldenis Squash Commit

Code Review LGTM :+1:

rousseldenis commented:
> @rousseldenis Squash Commit
> 
> Code Review LGTM

I'd prefer let them as it reflects development.

Cedric-Pigeon commented:
/ocabot merge nobump

rousseldenis commented:
/ocabot merge patch
